docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,474 @@
+.. _customizing-library-models-for-javascript:
+
+:orphan:
+:nosearch:
+
+Customizing Library Models for JavaScript
+=========================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+The JavaScript analysis can be customized by adding library models in data extension files.
+
+A data extension for JavaScript is a YAML file of form:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: <name of extension point>
+      data:
+        - <tuple1>
+        - <tuple2>
+        - ...
+
+The data extension can contribute to the following extension points:
+
+- **sourceModel**\(type, path, kind)
+- **sinkModel**\(type, path, kind)
+- **typeModel**\(type1, type2, path)
+- **summaryModel**\(type, path, input, output, kind)
+
+TODO: mention how to actually load the data extension and/or link to page about data extensions.
+
+We'll explain how to use these using a few examples, and provide some reference material at the end of this article.
+
+Example: Taint sink in the 'execa' package
+------------------------------------------
+
+In this example, we'll show how to add the argument passed to **execa** below as a command-line injection sink:
+
+.. code-block:: js
+
+  import { shell } from "execa";
+  shell(cmd); // <-- add 'cmd' as a taint sink
+
+This sink is already recognized by the CodeQL JS analysis, but for the sake of this example we'll show how it could be added as an extension.
+
+This can be achieved with the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["execa", "Member[shell].Argument[0]", "command-line-injection"]
+
+To break this down:
```

Comment:
```suggestion

```

Likely a bit too idiomatic, particularly for people for whom English isn't a first language; I think it makes sense without